javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -46,287 +45,410 @@ object Matcher {
-                                    var inputSequence: CharSequence)
+                                    var _inputSequence: CharSequence)
     extends MatchResult {
 
-  private def withRE2Regex[A](f: RE2RegExpOps => A): A =
-    _pattern.withRE2Regex(f)
+  private var currentPattern = _pattern
 
   private var hasMatch  = false
   private var hasGroups = false
   private var appendPos = 0
 
-  private var groups = withRE2Regex { re2op =>
-    val regex = re2op.ptr
-    Array.ofDim[(Int, Int)](cre2.numCapturingGroups(regex) + 1)
+  private var _regionStart = 0
+  private var _regionEnd   = _inputSequence.length
+
+  private var _groupCount = currentPattern.compiled.groupCount()
+
+  // This default differs from JVM but describes actual SN behavior.
+  // That behavior is a bug, awaiting a fix.
+  private var anchoringBoundsInUse = false
+
+  private def clearGroup0(g: Array[Int]) = {
+    g(0) = -1
+    g(1) = -1
   }
 
-  private var lastAnchor: Option[cre2.anchor_t] = None
+  private def createGroups(nGroups: Int): Array[Int] = {
 
-  private[regex] def inputLength = inputSequence.length
+    // +1 is for groups(0). It holds full match, but is not part of groupCount.
+    // "groups" is an old-style 2-D row-major array, where Array(index) holds
+    // start for group number index and Array(index + 1) holds the
+    // corresponding end. In SN, 1-D arrays are much more efficient, albeit
+    // much less elegant.
 
-  def matches(): Boolean = genMatch(0, ANCHOR_BOTH)
+    val result = new Array[Int](2 * (nGroups + 1))
 
-  def lookingAt(): Boolean = genMatch(0, ANCHOR_START)
+    // No lastmatch at creation, that is group(0) is coded as null,
+    // not null string.
+    // groups n > 0 have undefined values until after a loadGroup() executes.
+    clearGroup0(result)
 
-  def find(start: Int): Boolean = {
+    result
+  }
 
-    if ((start < 0) || (start > inputLength)) {
-      throw new IndexOutOfBoundsException("Illegal start index")
+  private def clearGroups(g: Array[Int]) = {
+    Arrays.fill(groups, 0, g.length, -1) // Now, all groups return null.
+  }
+
+  // The group indexes, in [start, end) pairs.  Zeroth pair is overall match.
+  private var groups = createGroups(_groupCount)
+
+  private def groupStartIndex(groupNum: Int) = groups(groupNum * 2)
+  private def groupEndIndex(groupNum: Int)   = groups((groupNum * 2) + 1)
+
+  private var lastAnchor: Option[RE2.anchor_t] = None
+
+  private[regex] def inputLength = _inputSequence.length
+
+  private[regex] def appendReplacement2(sb: StringBuffer,
+                                        replacement: String,
+                                        doGroups: Boolean): Matcher = {
+    val s = start
+    val e = end
+    if (appendPos < s) {
+      sb.append(_inputSequence, appendPos, s)
     }
+    appendPos = e
 
-    reset()
-    genMatch(start, UNANCHORED)
-  }
+    if (doGroups) {
+      val m =
+        Pattern.compile("(\\$(\\d)|\\$\\{(\\w*)\\})").matcher(replacement)
```

Comment:
Can you cache this `Pattern.compile` as a lazy val at top level somewhere?

Reply:
My most recent iteration of Matcher.scala does more delegation and is much simpler than
the iteration you looked at.  I believe the section you commented on no longer exists
in j.u.r.Matcher.scala.

I checked that no similar code exists in re2s.Matcher.scala.  The two are often parallel but
not here.